src/input/input_python.rs

```diff
@@ -910,3 +921,126 @@ impl<'py> KeywordArgs<'py> for PyKwargs<'py> {
         self.0.iter().map(Ok)
     }
 }
+
+#[cfg_attr(debug_assertions, derive(Debug))]
+pub enum GenericPyMapping<'a, 'py> {
+    Dict(&'a Bound<'py, PyDict>),
+    Mapping(&'a Bound<'py, PyMapping>),
+    GetAttr(Bound<'py, PyAny>, Option<Bound<'py, PyDict>>),
+}
+
+impl<'py> ValidatedDict<'py> for GenericPyMapping<'_, 'py> {
+    type Key<'a> = Bound<'py, PyAny>
+    where
+        Self: 'a;
+
+    type Item<'a> = Bound<'py, PyAny>
+    where
+        Self: 'a;
+
+    fn get_item<'k>(
+        &self,
+        key: &'k crate::lookup_key::LookupKey,
+    ) -> ValResult<Option<(&'k crate::lookup_key::LookupPath, Self::Item<'_>)>> {
+        match self {
+            Self::Dict(dict) => key.py_get_dict_item(dict),
+            Self::Mapping(mapping) => key.py_get_mapping_item(mapping),
+            Self::GetAttr(obj, dict) => key.py_get_attr(obj, dict.as_ref()),
+        }
+    }
+
+    fn as_py_dict(&self) -> Option<&Bound<'py, PyDict>> {
+        match self {
+            Self::Dict(dict) => Some(dict),
+            _ => None,
+        }
+    }
+
+    fn iterate<'a, R>(
+        &'a self,
+        consumer: impl ConsumeIterator<ValResult<(Self::Key<'a>, Self::Item<'a>)>, Output = R>,
+    ) -> ValResult<R> {
+        match self {
+            Self::Dict(dict) => Ok(consumer.consume_iterator(dict.iter().map(Ok))),
+            Self::Mapping(mapping) => Ok(consumer.consume_iterator(iterate_mapping_items(mapping)?)),
+            Self::GetAttr(obj, _) => Ok(consumer.consume_iterator(iterate_attributes(obj))),
+        }
+    }
+}
+
+fn iterate_mapping_items<'a, 'py>(
+    mapping: &'a Bound<'py, PyMapping>,
+) -> ValResult<impl Iterator<Item = ValResult<(Bound<'py, PyAny>, Bound<'py, PyAny>)>> + 'a> {
+    let py = mapping.py();
+    let input = mapping.as_any();
+    let iterator = mapping
+        .items()
+        .map_err(|e| mapping_err(e, py, input))?
+        .iter()
+        .map_err(|e| mapping_err(e, py, input))?
+        .map(move |item| match item {
+            Ok(item) => item.extract().map_err(|_| {
+                ValError::new(
+                    ErrorType::MappingType {
+                        error: MAPPING_TUPLE_ERROR.into(),
+                        context: None,
+                    },
+                    input,
+                )
+            }),
+            Err(e) => Err(mapping_err(e, py, input)),
+        });
+    Ok(iterator)
+}
+
+fn mapping_err<'py>(err: PyErr, py: Python<'py>, input: &'py (impl Input<'py> + ?Sized)) -> ValError {
+    ValError::new(
+        ErrorType::MappingType {
+            error: py_err_string(py, err).into(),
+            context: None,
+        },
+        input,
+    )
+}
+
+const MAPPING_TUPLE_ERROR: &str = "Mapping items must be tuples of (key, value) pairs";
+
+fn iterate_attributes<'a, 'py>(
```

Comment:
shame about the churn of moving all this from `return_enums.rs` is how hard would it be to put this there?

Reply:
Reverted; reduces the diff a little I guess.